File: PaymentSDKObjC/PaymentSDKObjC/MyBackendLib.m
//
//  MyBackendLib.m
//  PaymentSDKObjC
//
//  Created by Gustavo Sotelo on 27/11/17.
// Sample of  MyBackendLib. The implementation of these methods may be different on your application.
//  Copyright © 2017 Payment. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <PaymentSDK/PaymentSDK-Swift.h>
#import "MyBackendLib.h"

@implementation MyBackendLib
static NSString *myBackendUrl = @"https://example-Payment-backend.herokuapp.com/";

+(NSString *)myBackendUrl{
    return myBackendUrl;
}

+ (void) verifyTrxWithId:(NSString*)uid type:(int)type value:(NSString*)val callback:(void (^)(PaymentSDKError *error, BOOL *verified))callback
{
    //TODO implement verify sample
}
+ (void) deleteCardWithId:(NSString*)uid cardToken:(NSString*)cardToken callback:(void (^)(BOOL *deleted))callback
{
    //TODO implement deletecard 
}
+ (void) debitCardWithId:(NSString*)uid cardToken:(NSString*)cardToken sessionId:(NSString*)sessionId devReference:(NSString*)devReference amount:(double)amount description:(NSString*)description callback:(void (^)(NSError *error, PaymentTransaction *transaction))callback
{
    NSDictionary *params = @{@"uid": uid, @"token":cardToken, @"session_id":sessionId, @"amount": [NSString stringWithFormat:@"%.2f", amount], @"dev_reference":devReference, @"description":description};
    
    [self makeRequestPostFromUrl:@"create-charge" parameters:params responseCallback:^(NSError *error, NSInteger status, id responseData) {
        if (error == nil)
        {
            if (status == 200)
            {
                NSDictionary *responseD = (NSDictionary *)responseData;
                NSDictionary *transaction = responseD[@"transaction"];
                PaymentTransaction *trx = [[PaymentTransaction alloc] init];
                
                trx.amount = transaction[@"amount"];
                trx.status = transaction[@"status"];
                trx.statusDetail = transaction[@"status_detail"];
                trx.transactionId = transaction[@"id"];
                trx.carrierCode = transaction[@"carrier_code"];
                trx.message = transaction[@"message"];
                
                NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
                dateFormatter.dateFormat = @"yyyy-MM-dd'T'HH:mm:ssZZZZZ";
                
                if (responseD[@"payment_date"] != nil)
                {
                    trx.paymentDate = [dateFormatter dateFromString:responseD[@"payment_date"]];
                }
                if ([trx.statusDetail integerValue] != 3)
                {
                    callback(error,trx);
                }
                else
                {
                    callback(nil, trx);
                }
                
                
            }
            else
            {
              callback(error,nil);
            }
        }
        else
        {
            callback(error,nil);
        }
        
    }];
    
}
+ (void) listCardWithId:(NSString*)uid callback:(void (^)(NSMutableArray *cardList))callback
{
    NSDictionary *parameters = @{@"uid" : uid};
    
    [self makeRequestGetFromUrl:@"get-cards" parameters:parameters
               responseCallback:^(NSError *error, NSInteger status, id responseData) {
                   NSMutableArray *responseCards = [[NSMutableArray alloc] init];
                   if (error == nil)
                   {
                       if(status == 200)
                       {
                           NSDictionary *responseObj = (NSDictionary*)responseData;
                           NSArray *cardsArray = (NSArray*)responseObj[@"cards"];
                           for(NSDictionary* card in cardsArray)
                           {
                               PaymentCard *pCard = [[PaymentCard alloc] init];
                               pCard.bin = (NSString*) card[@"bin"];
                               pCard.token = (NSString*)  card[@"token"];
                               pCard.cardHolder = (NSString*)  card[@"holder_name"];
                               pCard.expiryMonth = (NSString*) card[@"expiry_month"];
                               pCard.expiryYear = (NSString*) card[@"expiry_year"];
                               pCard.termination = (NSString*)  card[@"number"];
                               pCard.status = (NSString*) card[@"status"];
                               pCard.type = (NSString*) card[@"type"];
                               [responseCards addObject:pCard];
                           }
                           
                           callback(responseCards);
                       }
                       else
                       {
                           callback(responseCards);
                       }
                   }
                   else
                   {
                       callback(responseCards);
                   }
               }];
    
}
+ (void) makeRequestGetFromUrl:(NSString*)urlToRequest parameters:(NSDictionary*)parameters responseCallback:(void (^)(NSError *error, NSInteger status,  id responseData))responseCallback
{
    NSLog(@"%@",self.myBackendUrl);
    NSString *completeUrl = [self.myBackendUrl stringByAppendingString:urlToRequest];
    NSString *bodyData = @"";
    for (NSString* key in parameters) {
        NSString *value = (NSString*)[parameters objectForKey:key];
        NSString *scapedKey = [key stringByAddingPercentEncodingWithAllowedCharacters:NSCharacterSet.URLHostAllowedCharacterSet];
        NSString *scapedValue = [value stringByAddingPercentEncodingWithAllowedCharacters:NSCharacterSet.URLHostAllowedCharacterSet];
        bodyData = [NSString stringWithFormat:@"%@&%@=%@", bodyData, scapedKey, scapedValue];
        // do stuff
    }
    completeUrl = [NSString stringWithFormat:@"%@?%@", completeUrl, bodyData];
    NSURL *url = [NSURL URLWithString:completeUrl];
    
    NSURLSession *session = NSURLSession.sharedSession;
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = @"GET";
    
    NSURLSessionDataTask *task = [session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        NSLog(@"%@", data);
        if (error == nil)
        {
            id json = nil;
            
            json = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableLeaves error:&error];
            
            if (!json) {
                responseCallback(error,400,nil);
            } else {
                NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *) response;
                responseCallback(error, [httpResponse statusCode], json);
            }
        }
        else
        {
            responseCallback(error, 400, nil);
        }
    }];
    [task resume];
}
+ (void) makeRequestPostFromUrl:(NSString*)urlToRequest parameters:(NSDictionary*)parameters responseCallback:(void (^)(NSError *error, NSInteger status,  id responseData))responseCallback
{
    NSString *completeUrl = [self.myBackendUrl stringByAppendingString:urlToRequest];
    NSURL *url = [NSURL URLWithString:completeUrl];
    
    NSURLSession *session = NSURLSession.sharedSession;
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = @"POST";
    request.HTTPBody = [self encodeParameters:parameters];
    NSURLSessionDataTask *task = [session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error == nil)
        {
            id json = nil;
            
            json = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableLeaves error:&error];
            
            if (!json) {
                responseCallback(error,400,nil);
            } else {
                NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *) response;
                responseCallback(error, [httpResponse statusCode], json);
            }
        }
        else
        {
            responseCallback(error, 400, nil);
        }
    }];
    [task resume];
    
}

+ (NSData*) encodeParameters:(NSDictionary*)parameters
{
    NSString *bodyData = @"";
    for (NSString* key in parameters) {
        NSString *value = (NSString*)[parameters objectForKey:key];
        NSString *scapedKey = [key stringByAddingPercentEncodingWithAllowedCharacters:NSCharacterSet.URLHostAllowedCharacterSet];
        NSString *scapedValue = [value stringByAddingPercentEncodingWithAllowedCharacters:NSCharacterSet.URLHostAllowedCharacterSet];
        bodyData = [NSString stringWithFormat:@"%@&%@=%@", bodyData, scapedKey, scapedValue];
        // do stuff
    }
    return [bodyData dataUsingEncoding:NSUTF8StringEncoding];
}



@end

File: PaymentSDKSwift/PaymentSDKSwift/MyBackendLib.swift
    //
    //  MyBackendLib.swift
    //  PaymentSDKSwift
    //
    //  Created by Gustavo Sotelo on 19/09/17.
    //  Copyright © 2017 Payment. All rights reserved.
    //
    
    import Foundation
    import PaymentSDK
    
    class MyBackendLib
    {
        static let myBackendUrl = "https://example-Payment-backend.herokuapp.com/"
        
        static func verifyTrx(uid:String, transactionId:String, type:Int, value:String, callback:@escaping (_ error:PaymentSDKError?, _ verified:Bool) ->Void)
        {
            var verifyType = "BY_AMOUNT"
            if type == 1
            {
                verifyType = "BY_AUTH_CODE"
            }
            
            let params = ["uid": uid, "transaction_id": transactionId, "type": verifyType, "value":value]
            makeRequestPOST("verify-transaction", parameters: params as NSDictionary) { (error, statusCode, responseData) in
                
                if statusCode == 200
                {
                    callback(nil,true)
                }
                else
                {
                    do {
                        var dataR = (responseData as! [String:Any])
                        if dataR["error"] != nil
                        {
                            dataR = dataR["error"] as! [String:Any]
                        }
                        let errorPa =  PaymentSDKError.createError(statusCode!, description: dataR["description"] as! String, help: dataR["help"] as? String, type: dataR["type"] as? String) //parse error in strutcture
                        
                        callback(errorPa,false)
                    }
                }
            }
        }
        
        
        static func deleteCard(uid:String, cardToken:String, callback:@escaping (_ deleted:Bool) ->Void)
        {
            let params = ["token": cardToken, "uid": uid]
            makeRequestPOST("delete-card", parameters: params as NSDictionary) { (error, statusCode, responseData) in
                
                if statusCode == 200
                {
                    callback(true)
                }
                else
                {
                    callback(false)
                }
            }
        }
        
        
        static func debitCard(uid:String, cardToken:String, sessionId:String, devReference:String, amount:Double, description:String, callback:@escaping (_ error:NSError?, _ transaction:PaymentTransaction?) ->Void)
        {
            let params = ["uid": uid, "token": cardToken, "session_id": sessionId, "amount": String(format: "%.2f", amount), "dev_reference":devReference, "description":description] as NSDictionary
            makeRequestPOST("create-charge", parameters: params) { (error, statusCode, responseData) in
                if error == nil
                {
                    if statusCode == 200
                    {
                        let responseD = responseData as! [String:Any]
                        let transaction = responseD["transaction"] as! [String:Any]
                        let trx = PaymentTransaction()
                        trx.amount = transaction["amount"] as? NSNumber
                        trx.status = transaction["status"] as? String
                        trx.statusDetail = transaction["status_detail"] as? Int as NSNumber?
                        trx.transactionId = transaction["id"] as? String
                        if let authCode = Int(transaction["authorization_code"] as! String) {
                            trx.authorizationCode = NSNumber(value:authCode)
                        }
                            
                        trx.carrierCode = transaction["carrier_code"] as? String
                        trx.message = transaction["message"] as? String
                        
                        
                        let dateFormatter = DateFormatter()
                        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZZZZZ"
                        
                        if let paymentdate = (responseD["payment_date"] as? String)
                        {
                            trx.paymentDate = dateFormatter.date(from: paymentdate)
                        }
                        if trx.statusDetail == 11
                        {
                            callback(nil, trx)
                        }
                        else if trx.statusDetail != 3
                        {
                            
                            callback(error, trx)
                        }
                        else
                        {
                            callback(nil, trx)
                        }
                        
                        
                    }
                    else
                    {
                        do {
                       
                            
                            callback(error, nil)
                        }
                        
                    }
                }
                else
                {
                    callback(error, nil)
                }
            }
        }
        static func listCard(uid:String, callback:@escaping (_ cardList:[PaymentCard]?) ->Void)
        {
            let parameters = ["uid": uid] as [String:Any]
            makeRequestGet("get-cards", parameters: parameters as NSDictionary) { (error, statusCode, responseData) in
                
                
                var responseCards = [PaymentCard]()
                if error == nil
                {
                    if statusCode == 200
                    {
                        let responseObj = responseData as! NSDictionary
                        let cardsArray = responseObj["cards"] as! [[String:Any]]
                        
                        
                        for card in cardsArray
                        {
                            let pCard = PaymentCard()
                            pCard.bin = "\(card["bin"]!)"
                            pCard.token = "\(card["token"]!)"
                            pCard.cardHolder = "\(card["holder_name"]!)"
                            pCard.expiryMonth = card["expiry_month"] as! String
                            pCard.expiryYear = card["expiry_year"] as! String
                            pCard.termination = "\(card["number"]!)"
                            pCard.status = "\(card["status"]!)"
                            pCard.type = "\(card["type"]!)"
                            responseCards.append(pCard)
                        }
                    }
                    
                    callback(responseCards)
                        
                    
                }
                else
                {
                    callback(responseCards)
                }
                
            }
        }
        static func makeRequestGet(_ urlToRequest:String, parameters:NSDictionary, responseCallback:@escaping (_ error:NSError?, _ statusCode:Int?,_ responseData:Any?) ->Void)
        {
            var completeUrl = self.myBackendUrl + urlToRequest
            var bodyData = ""
            for (key,value) in parameters{
                let scapedKey = (key as! String).addingPercentEncoding(
                    withAllowedCharacters: .urlHostAllowed)!
                let scapedValue = (value as! String).addingPercentEncoding(
                    withAllowedCharacters: .urlHostAllowed)!
                bodyData += "\(scapedKey)=\(scapedValue)&"
            }
            completeUrl += "?" + bodyData
            let url:URL? = URL(string: completeUrl)
            let session = URLSession.shared
            
            var request = URLRequest(url:url!)
            request.httpMethod = "GET"
            
            let task = session.dataTask(with: request){ data, resp, err in
                
                if err == nil
                {
                    var json:Any? = nil
                    
                    do{
                        json = try JSONSerialization.jsonObject(with: data!, options: .mutableLeaves)
                    }
                    catch {
                        
                    }
                    
                    if json == nil{
                        let error = NSError(domain: "JSON", code: 400, userInfo: ["parsing" : false])
                        print ("parsing json error")
                        responseCallback(error, 400,nil)
                    }
                    else
                    {
                        
                        
                        responseCallback(err as NSError?, (resp as! HTTPURLResponse).statusCode, json)
                    }
                }
                else
                {
                    responseCallback(err as NSError?, 400, nil)
                }
            }
            task.resume()
        }
        static func makeRequestPOST(_ urlToRequest:String, parameters:NSDictionary, responseCallback:@escaping (_ error:NSError?, _ statusCode:Int?,_ responseData:Any?) ->Void)
        {
            let completeUrl = self.myBackendUrl + urlToRequest
            let url:URL? = URL(string: completeUrl)
            let session = URLSession.shared
            
            var request = URLRequest(url:url!)
            request.httpMethod = "POST"
            request.httpBody = encodeParams(parameters)
            
            let task = session.dataTask(with: request){
                data, resp, err in
                print(data as Any)
                if err == nil
                {
                    var json:Any? = nil
                    
                    do{
                        json = try JSONSerialization.jsonObject(with: data!, options: .mutableLeaves)
                        print(json as Any)
                    }
                    catch {
                        
                    }
                    
                    if json == nil{
                        let error = NSError(domain: "JSON", code: 400, userInfo: ["parsing" : false])
                        print ("parsing json error")
                        if resp != nil
                        {
                            
                            responseCallback(error, (resp as! HTTPURLResponse).statusCode,nil)
                        }
                        else
                        {
                            responseCallback(error, 400,nil)
                        }
                    }
                    else
                    {
                        
                        print(json as Any)
                        responseCallback(err as NSError?, (resp as! HTTPURLResponse).statusCode, json)
                    }
                }
                else
                {
                    responseCallback(err as NSError?, 400, nil)
                }
            }
            task.resume()
        }
        static func encodeParams(_ parameters:NSDictionary) ->Data?
        {
            var bodyData = ""
            for (key,value) in parameters{
                let scapedKey = (key as! String).addingPercentEncoding(
                    withAllowedCharacters: .urlHostAllowed)!
                let scapedValue = (value as! String).addingPercentEncoding(
                    withAllowedCharacters: .urlHostAllowed)!
                bodyData += "\(scapedKey)=\(scapedValue)&"
            }
            print(bodyData)
            return bodyData.data(using: String.Encoding.utf8, allowLossyConversion: true)
        }
    }
    
